pkg/reconciler/apis/apibindingdeletion/apibinding_deletion_controller.go

```diff
@@ -0,0 +1,359 @@
+		)
+
+		return apibinding, &deletion.ResourcesRemainingError{Estimate: DeletionRecheckEstimateSeconds}
+	}
+
+	if len(resourceRemaining.gvrToNumRemaining) != 0 {
+		// requeue if there are still remaining resources
+		remainingResources := []string{}
+		for gvr, numRemaining := range resourceRemaining.gvrToNumRemaining {
+			if numRemaining == 0 {
+				continue
+			}
+			remainingResources = append(remainingResources, fmt.Sprintf("%s.%s has %d resource instances", gvr.Resource, gvr.Group, numRemaining))
+		}
+		// sort for stable updates
+		sort.Strings(remainingResources)
+
+		conditions.MarkFalse(
+			apibinding,
+			apisv1alpha1.BindingResourceDeleteSuccess,
+			ResourceRemainingReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			fmt.Sprintf("Some resources are remaining: %s", strings.Join(remainingResources, ", ")),
+		)
+
+		return apibinding, &deletion.ResourcesRemainingError{Estimate: DeletionRecheckEstimateSeconds}
+	}
+
+	conditions.MarkTrue(apibinding, apisv1alpha1.BindingResourceDeleteSuccess)
+
+	return apibinding, nil
+}
+
+func (c *Controller) patchCondition(ctx context.Context, old, new *apisv1alpha1.APIBinding) error {
+	if equality.Semantic.DeepEqual(old.Status.Conditions, new.Status.Conditions) {
+		return nil
+	}
+
+	oldData, err := json.Marshal(apisv1alpha1.APIBinding{
+		Status: apisv1alpha1.APIBindingStatus{
+			Conditions: old.Status.Conditions,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to Marshal old data for apibinding %s: %w", old.Name, err)
+	}
+
+	newData, err := json.Marshal(apisv1alpha1.APIBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			UID:             old.UID,
+			ResourceVersion: old.ResourceVersion,
+		}, // to ensure they appear in the patch as preconditions
+		Status: apisv1alpha1.APIBindingStatus{
+			Conditions: new.Status.Conditions,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to Marshal new data for apibinding %s: %w", new.Name, err)
+	}
+
+	patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+	if err != nil {
+		return fmt.Errorf("failed to create patch for apibinding %s: %w", new.Name, err)
+	}
+
+	_, err = c.kcpClusterClient.Cluster(logicalcluster.From(new)).ApisV1alpha1().APIBindings().Patch(ctx, new.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	return err
+}
+
+// finalizeAPIBinding removes the specified finalizer and finalizes the apibinding
+func (c *Controller) finalizeAPIBinding(ctx context.Context, apibinding *apisv1alpha1.APIBinding) error {
+	copiedFinalizers := []string{}
+	for i := range apibinding.Finalizers {
+		if apibinding.Finalizers[i] == APIBindingFinalizer {
+			continue
+		}
+		copiedFinalizers = append(copiedFinalizers, APIBindingFinalizer)
+	}
+	if len(apibinding.Finalizers) == len(copiedFinalizers) {
+		return nil
+	}
+
+	finalizerBytes, err := json.Marshal(copiedFinalizers)
+	if err != nil {
+		return err
+	}
+	patch := fmt.Sprintf("{\"metadata\": {\"finalizers\": %s}}", string(finalizerBytes))
```

Comment:
OK, then perhaps two cases:

1. finalizers is empty -> use a static, back-ticks string
2. not empty -> use JSON Patch